V1 Automation

anttix opened the pull request.

Commit:
Add V1 workflows

Commit:
Add SKR 1.3 dual endstop version

jeffeb3 commented on .github/workflows/v1-builds.yml:
```diff
@@ -0,0 +1,132 @@
+      run: |
+        v1-scripts/fetch-branch "origin/$BUILD_BRANCH"
+        git fetch --prune origin "$MIRROR_BRANCH" # We need full history of the mirror branch
+        git branch $MIRROR_BRANCH origin/$MIRROR_BRANCH
```
I might not be getting the order here, but should this be fetching from upstream?

jeffeb3 replied:
N/M, I see the pull action at 15 */6

jeffeb3 commented on .github/workflows/v1-builds.yml:
```diff
@@ -0,0 +1,132 @@
+    - name: Push
+      if: github.event_name != 'pull_request'
+      run: |
+        git push origin "$BUILD_BRANCH:refs/heads/$BUILD_BRANCH" --force
```
Maybe I don't understand what this is doing...

jeffeb3 commented on .github/workflows/v1-builds.yml:
```diff
@@ -0,0 +1,132 @@
+        #bodyFile: "body.md"
+        token: ${{ secrets.GITHUB_TOKEN }}
+        prerelease: true
+        name: unstable
```
So this will always go to the same release, right? There won't be a way to get a previous unstable branch?

jeffeb3 replied:
Are these build products also available in the build log?

anttix replied:
I'm not too happy with existing release plugins. It starts to look like we may have to build one.

In principle one can upload artifacts from build steps that can then be downloaded from the steps. But this is difficult to find/use and it's not designed for long-term storage.

If there is a need to keep a longer history of nightlies these must be published as releases. Which means there must be some cleanup logic to prune these.

jeffeb3 commented on v1-scripts/pull-from-upstream:
```diff
@@ -0,0 +1,35 @@
+MIRROR_BRANCH="${MIRROR_BRANCH:-$UPSTREAM_BRANCH}"
+ORIGIN_REMOTE_NAME="${ORIGIN_REMOTE_NAME:-origin}"
+
+PULL_REQUESTS="${PULL_REQUESTS:-16108 16229}"
```
Interesting.

jeffeb3 replied:
These have been merged, so we can get rid of these, right?

jeffeb3 replied:
This will also only work for PRs against bugfix, otherwise, it will merge stuff not in the PR.

anttix replied:
Yeah, these are not needed any more. I keep these around to test the concept.

jeffeb3 commented on v1-scripts/pull-from-upstream:
```diff
@@ -0,0 +1,35 @@
+
+UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY:-https://github.com/MarlinFirmware/Marlin.git}"
+UPSTREAM_REMOTE_NAME="${UPSTREAM_REMOTE_NAME:-upstream}"
+UPSTREAM_BRANCH="${UPSTREAM_BRANCH:-$UPSTREAM_REMOTE_NAME/bugfix-2.0.x}"
```
This could be an argument, to support picking this branch from upstream. I think 2.0.x is probably more useful for us, but bugfix is also useful.

jeffeb3 replied:
Oh, it is. Because of the '=-'. Got it.

jeffeb3 commented on v1-scripts/prepare-release:
```diff
@@ -0,0 +1,16 @@
+
+mkdir -p .zip
+ln -sf ".." ".zip/$dirname-src"
+(cd ".zip" && zip -prq "$flatname-src.zip" "$dirname-src"/*)
```
Do we need to clean up any build products before zipping?

anttix replied:
The magic of /* - it will not pick up .pio - the implication is that build artifacts should always be stored in directories that start with a dot.

jeffeb3 commented on v1-scripts/generate-configs:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
I love this step!

jeffeb3 commented on v1-scripts/generate-build-branch:
```diff
@@ -0,0 +1,33 @@
+MIRROR_BRANCH="${MIRROR_BRANCH:-bugfix-2.0.x}"
+MACHINES_BRANCH="${MACHINES_BRANCH:-HEAD}"
+BUILD_BRANCH="${BUILD_BRANCH:-v1-$MIRROR_BRANCH}"
+ADD_BRANCHES="${ADD_BRANCHES:-v1-base-config}"
```
This v1-base-config has to always be based on a common parent, or it will pull inextra stuff, right? I still have this idea in my head that bugfix and releases are not organized well upstream. Not sure how to fix that though.

jeffeb3 replied:
What if v1-base-config was a submodule (or even just a nested checkout during build).

anttix replied:
The point of these branches is to carry code that needs to be merged in but upstream will not carry. Think about backports of our PR-s that make it work/build on 2.0.x. These are present in bugfix-2.0.x and will be eventually in 2.0.1 but until that happens, backports need to be stored somewhere.

The v1-base-config branch right now is a test of two concepts: Starting with a "base" config that we (hopefully) push to upsteam and this branch merging logic.

jeffeb3 commented on v1-scripts/configs/ZenXY/Ramps_16T_LCD_32:
```diff
@@ -0,0 +1,11 @@
+. $incdir/drv8825
+
+opt_set DEFAULT_AXIS_STEPS_PER_UNIT "{ 200, 200, 800, 200 }"
+opt_enable REVERSE_ENCODER_DIRECTION
```
I think reverse encoder direction should probably be the same on all the LCD configs. I'm guessing there is an error in the configs that causes this to only be on some setups.

jeffeb3 replied:
This probably should be put next to the disolay enable feature, which looks like it is in the base configs.

jeffeb3 commented on v1-scripts/configs/MPCNC/SKR1.3_T8_16T_LCD_32step_DualEndstop:
```diff
@@ -0,0 +1,15 @@
+
+incdir=$(dirname $0)/..
+
+use_example_configs CNC/MPCNC
```
You don't want this, do you? This is going to have a chicken and egg problem.

jeffeb3 replied:
Or is this just copying the _Bootscreen?

anttix replied:
This copies the entire config/examples/CNC/MPCNC directory.

jeffeb3 commented on v1-scripts/configs/MPCNC/Rambo_T8_16T_LCD_DualEndstop:
```diff
@@ -0,0 +1,15 @@
+
+opt_enable REVERSE_ENCODER_DIRECTION
+
+# Why?
```
No idea. Probably just not as obvious as it is here.

jeffeb3 reviewed (approved):
I am half distracted ATM. I want to check this out on my computer and poke around with it. I'm really not sure how, for example, we would make the releases for the latest 2.0.1 release. 

I am super happy about this though. This is going to really increase the quality.

Commit:
Make config generation errors fatal

Commit:
Fix local usage of v1-scripts/pull-from-upstream

Commit:
Fix PIO environment for Rambo boards

Commit:
Move Marlin to a git sub-module

Commit:
Add a dual endstop config for SKR Pro 1.1

Commit:
Add configs for 2.0.x and dev-2.1.x branches

dolfelt committed:
Add readme and provide double check in script

Commit:
Enable helical plunge with arcs (P circles) and CNC workplanes

Commit:
Correct stepper type for Archim2 board

V1EngineeringInc committed:
My current Base config

I have no idea if I did this right but I am getting closer to finishing up all the new firmware and these are the settings I am using.

Trying to keep my note in a public place, work in progress https://docs.google.com/spreadsheets/d/1vqvdtS2TFGz9YddYO6rAU1w7-0tvp0B9ObIfAOfCKvs/edit?usp=sharing

Commit:
Remove already merged PRs from bugfix-2.0.x build

Commit:
Ditch the hand-edited base config and switch to 100% generated

Commit:
Fix build error with latest upstream (build_marlin_pio script was removed)

Commit:
Delete upstream workflows on pull

Github doesn't allow automated pushes when workflows are present.

! [remote rejected]     HEAD -> v1-bugfix-2.0.x (refusing to allow a bot to create or update workflow `.github/workflows/test-builds.yml`)

Commit:
Rename dual-endstops to dual-drivers-on-xy

Commit:
Added single-axis config for SKR1.3

Commit:
Handle fallout from upstream's config repository split

Commit:
Update BTT board names

Commit:
Added SKR1.3 configuration for Lowrider dual driver setup

Commit:
Add TMC2209 support and sensorless config for SKR1.3 lowrider

Commit:
Rename megaatmega2560 to mega2560 to accomodate upstream changes

Commit:
Add support for TMC5160 drivers

Commit:
Publish an STL for Lowrider2 sensorless Z- homing

Commit:
Publish Coupling saver STL
